File: src/bin/0003-longest-substring-without-repeating-characters.py
# @leetup=info id=3 lang=python3 slug=longest-substring-without-repeating-characters

# @leetup=code


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        prev = {}
        start = 0
        result = 0
        for i, c in enumerate(s):
            if c in prev:
                start = max(start, prev[c] + 1)
            prev[c] = i
            result = max(result, i + 1 - start)
        return result


# @leetup=code

File: src/bin/0003-longest-substring-without-repeating-characters.cpp
// @leetup=info id=3 lang=cpp
// slug=longest-substring-without-repeating-characters

#include <algorithm>
#include <iostream>
#include <map>
#include <set>
#include <tuple>
#include <unordered_map>
#include <vector>

// @leetup=code

class Solution {
 public:
  int lengthOfLongestSubstring(std::string s) {
    std::unordered_map<char, int> prev;
    int left = 0;
    int result = 0;
    for (int i = 0; i < s.size(); ++i) {
      char c = s[i];
      auto found = prev.find(c);
      if (found != prev.end()) {
        left = std::max(left, found->second + 1);
        found->second = i;
      } else {
        prev[c] = i;
      }
      result = std::max(result, i + 1 - left);
    }
    return result;
  }
};
// @leetup=code

int main() {
  std::cout << "Hello, world!" << std::endl;
  return 0;
}
